### PR TITLE
Provide integration tests via the key Ditto features - Metrics

### PR DESCRIPTION
**Pull request by antonia-avramova**
[#71] Provide integration tests via the key Ditto features - Metrics

Signed-off-by: Antonia Avramova <antonia.avramova@bosch.io>

### CODE REVIEW

File: integration/ctr_metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2022 Contributors to the Eclipse Foundation
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:1]:
```suggestion
// Copyright (c) 2023 Contributors to the Eclipse Foundation
```

File: integration/ctr_metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+				return true, fmt.Errorf("Invalid metrics ID: %s", mm.ID)
+			}
+		}
+
+		return len(expectedOriginators) == len(actualOriginators), nil
+	})
+
+	return result
+}
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:235]: please refactor it something like this:
```suggestion
				if !allowedPrefixID(mm.ID) {
					return true, fmt.Errorf("Invalid metrics ID: %s", mm.ID)
				}
			}
		}

		return len(expectedOriginators) == len(actualOriginators), nil
	})

	return result
}

func allowedPrefixID(id string) bool {
	allowedIDs := []string{"cpu.", "memory.", "io.", "net.", "pids"}
	for _, allowedID := range allowedIDs {
		if strings.HasPrefix(id, allowedID){
			return true
		}
	}
	return false
}
```

**Review by k-gostev — CHANGES_REQUESTED** [review]

File: integration/ctr_metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+func (suite *ctrMetricsSuite) SetupSuite() {
+	suite.SetupCtrManagementSuite()
+	suite.metricsUrl = util.GetFeatureURL(suite.ctrThingURL, metricsFeatureID)
+
+	suite.pathData = util.GetFeatureOutboxMessagePath(metricsFeatureID, actionData)
+	suite.topicData = util.GetLiveMessageTopic(suite.ctrThingID, protocol.TopicAction(actionData))
+	suite.assertContainerMetricsFeatures()
+	params := make(map[string]interface{})
+	params[paramImageRef] = "docker.io/library/httpd:latest"
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:67]: `"docker.io/library/httpd:latest"` appears also [here](https://github.com/eclipse-kanto/container-management/blob/31ec917e881a090449e8aa61534a06acf18de217/integration/ctr_factory_test.go#L94). Please move it to `car_management_test_constants.go`

File: integration/ctr_metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	metricsUrl         string
+	pathData           string
+	topicData          string
+}
+
+var (
+	stopParams = map[string]interface{}{
+		paramFrequency: "0s",
+	}
+)
```
Comment (antoniyatrifonova) [inline integration/ctr_metrics_test.go:57]: You don't need this constants. You the value directly.

**Review by antoniyatrifonova — CHANGES_REQUESTED** [review]

File: integration/ctr_metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	"testing"
+	"time"
+
+	"github.com/eclipse-kanto/container-management/containerm/things"
+	"github.com/eclipse-kanto/kanto/integration/util"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	// "golang.org/x/net/websocket"
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:31]:
```suggestion
```

File: integration/ctr_metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	params[paramFilter] = []things.Filter{filter}
+
+	err := suite.testMetrics(params, filter.Originator)
+	assert.Errorf(suite.T(), err,
+		"metrics event for non existing originator '%s' should not be received", filter.Originator)
+
+	filter.ID = []string{"test.io", "test.cpu", "test.memory", "test.net"}
+	filter.Originator = suite.secondCtrID
+	params[paramFrequency] = "1s"
+	params[paramFilter] = []things.Filter{filter}
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:109]: Why do we need to redefine the frequency? Can't we use the `2s` interval defined above?

File: integration/ctr_metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+	err := suite.testMetrics(params, suite.firstCtrID)
+	require.NoErrorf(suite.T(), err, "error while receiving metrics for '%s' container", suite.firstCtrID)
+}
+
+func (suite *ctrMetricsSuite) TestFilterNotMatching() {
+	params := make(map[string]interface{})
+	params[paramFrequency] = "2s"
+	filter := things.Filter{}
+	filter.Originator = "test.process"
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:99]:
```suggestion
filter := things.Filter{Originator: "test.process"}
```

File: integration/ctr_management_suite.go [hunk 8]
```diff
@@ -75,9 +75,16 @@ func (suite *ctrManagementSuite) createWSConnection() *websocket.Conn {
...
-func (suite *ctrManagementSuite) createOperation(operation string, params map[string]interface{}) (*websocket.Conn, string) {
+func (suite *ctrManagementSuite) createOperation(operation string, params map[string]interface{}) string {
 	wsConnection := suite.createWSConnection()
 
+	defer func() {
+		if wsConnection != nil {
+			util.UnsubscribeFromWSMessages(suite.Cfg, wsConnection, util.StopSendEvents)
+			wsConnection.Close()
+		}
+	}()
```
Comment (k-gostev) [inline integration/ctr_management_suite.go:86]: Please create separate function for this to avoid duplication.

**Review by k-gostev — CHANGES_REQUESTED** [review]

File: integration/ctr_metrics_test.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+	ctrParams := map[string]interface{}{
+		paramImageRef: httpdImageRef,
+		paramStart:    true,
+	}
+
+	suite.firstCtrID = suite.create(ctrParams)
+	suite.secondCtrID = suite.create(ctrParams)
+
+	suite.assertContainerMetricsFeatures()
```
Comment (antoniyatrifonova) [inline integration/ctr_metrics_test.go:60]: First assert the Metrics feature, before creating the containers. In the assertion verify that the definition of the feature is expected.

File: integration/ctr_metrics_test.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+func (suite *ctrMetricsSuite) TearDownSuite() {
+	suite.stopMetricsRequest()
```
Comment (antoniyatrifonova) [inline integration/ctr_metrics_test.go:64]: In the `testMetrics` you stop the metric request. Also in this method the `stopMetricRequest` is invoked in defer function, that means this method call will be executed in any case. So there is no need to invoke the method again.

File: integration/ctr_metrics_test.go [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+	params[paramFrequency] = "3s"
+
+	err := suite.testMetrics(params, suite.firstCtrID, suite.secondCtrID)
+	require.NoError(suite.T(), err, "error while receiving metrics for all container")
+}
+
+func (suite *ctrMetricsSuite) TestRequestMetricsForFirstContainer() {
+	filter := things.Filter{}
+	filter.ID = []string{"cpu.*", "memory.*", "io.*", "net.*", "pids"}
+	filter.Originator = suite.firstCtrID
```
Comment (antoniyatrifonova) [inline integration/ctr_metrics_test.go:85]:
```suggestion
		filter := things.Filter{
		ID: []string{"cpu.*", "memory.*", "io.*", "net.*", "pids"},
		Originator: suite.firstCtrID,
	}

File: integration/ctr_metrics_test.go [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+	assert.Errorf(suite.T(), err, "error while sending metrics request with invalid params %v", params)
+
+	params[paramFrequency] = "2s"
+	invalidAction := "invalidRequest"
+	_, err = util.ExecuteOperation(suite.Cfg, suite.metricsUrl, invalidAction, params)
+	assert.Errorf(suite.T(), err, "error while sending metrics request with wrong topic %v", invalidAction)
+}
+
+func (suite *ctrMetricsSuite) assertContainerMetricsFeatures() {
+	_, err := util.SendDigitalTwinRequest(suite.Cfg, http.MethodGet, suite.metricsUrl, nil)
```
Comment (antoniyatrifonova) [inline integration/ctr_metrics_test.go:125]: Assert for feature definition also.

**Review by antoniyatrifonova — CHANGES_REQUESTED** [review]

File: integration/ctr_factory_test.go [hunk 13]
```diff
@@ -45,19 +44,13 @@ func (suite *ctrFactorySuite) TestCreate() {
 	params[paramImageRef] = influxdbImageRef
 	params[paramStart] = true
 
-	var (
-		ctrFeatureID string
-		wsConnection *websocket.Conn
-	)
+	var ctrFeatureID string
 
 	defer func() {
```
Comment (k-gostev) [inline integration/ctr_factory_test.go:49]:
```suggestion
defer suite.remove(ctrFeatureID)
```

File: integration/ctr_metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+	require.NoError(suite.T(), err, "error while getting the metrics feature")
+}
+
+func (suite *ctrMetricsSuite) stopMetricsRequest() {
+	stopParams := map[string]interface{}{
+		paramFrequency: "0s",
+	}
+	suite.executeMetrics(stopParams)
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:133]:
```suggestion
	suite.executeMetrics(map[string]interface{}{paramFrequency: "0s"})
```

File: integration/ctr_metrics_test.go [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+	err = util.SubscribeForWSMessages(suite.Cfg, wsConnection, util.StartSendMessages, "")
+	defer func() {
+		suite.stopMetricsRequest()
+		util.UnsubscribeFromWSMessages(suite.Cfg, wsConnection, util.StopSendMessages)
+	}()
+	require.NoError(suite.T(), err, "unable to listen for events by using a websocket connection")
+
+	pathData := util.GetFeatureOutboxMessagePath(metricsFeatureID, actionData)
+	topicData := util.GetLiveMessageTopic(suite.ctrThingID, protocol.TopicAction(actionData))
```
Comment (k-gostev) [inline integration/ctr_metrics_test.go:154]: I think those two variables are used only once in the `if` statements below. Could you check if we can just use the util functions there and avoid declaring them?

**Review by k-gostev — CHANGES_REQUESTED** [review]
Just a couple of general suggestions.

**Review by antoniyatrifonova — APPROVED** [review]

**Review by k-gostev — APPROVED** [review]

File: integration/ctr_metrics_test.go [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+
+	suite.assertContainerMetricsFeatures()
+
+	ctrParams := map[string]interface{}{
+		paramImageRef: httpdImageRef,
+		paramStart:    true,
+	}
+
+	suite.firstCtrID = suite.create(ctrParams)
+	suite.secondCtrID = suite.create(ctrParams)
```
Comment (dimitar-dimitrow) [inline integration/ctr_metrics_test.go:60]: Those are actually the feature identifiers of the containers. first**CtrID** sounds like this is the identifier of the container and mislead me that the tests will fail as the metrics feature would not work with container id as an originator. Could you please change them to something like firstCtrFeatureID?

**Review by dimitar-dimitrow — CHANGES_REQUESTED** [review]

**Review by dimitar-dimitrow — APPROVED** [review]

**Pull request merged**